src/include/storage/copier/rel_copier.h

```diff
@@ -0,0 +1,278 @@
+
+    void execute(processor::ExecutionContext* executionContext);
+
+    virtual std::unique_ptr<RelCopier> clone() const {
+        return std::make_unique<RelCopier>(
+            sharedState, copyDesc, schema, fwdRelData, bwdRelData, pkColumnTypes, pkIndexes);
+    }
+    virtual void finalize() {
+        // DO NOTHING.
+    }
+
+    inline std::shared_ptr<RelCopySharedState> getSharedState() const { return sharedState; }
+
+protected:
+    virtual void executeInternal(std::unique_ptr<RelCopyMorsel> morsel) {
+        throw common::CopyException("RelCopier::executeInternal not implemented");
+    }
+
+    static void indexLookup(arrow::Array* pkArray, const common::LogicalType& pkColumnType,
+        PrimaryKeyIndex* pkIndex, common::offset_t* offsets);
+
+    inline void copyRelColumnsOrCountRelListsSize(common::offset_t tupleIdx,
+        arrow::RecordBatch* recordBatch, common::RelDataDirection direction,
+        const std::vector<std::unique_ptr<arrow::Array>>& pkOffsets) {
+        if (schema->isSingleMultiplicityInDirection(direction)) {
+            copyRelColumns(tupleIdx, recordBatch, direction, pkOffsets);
+        } else {
+            countRelListsSize(direction, pkOffsets);
+        }
+    }
+
+    void copyRelColumns(common::offset_t tupleIdx, arrow::RecordBatch* recordBatch,
+        common::RelDataDirection direction,
+        const std::vector<std::unique_ptr<arrow::Array>>& pkOffsets);
+    void countRelListsSize(common::RelDataDirection direction,
+        const std::vector<std::unique_ptr<arrow::Array>>& pkOffsets);
+    void copyRelLists(common::offset_t tupleIdx, arrow::RecordBatch* recordBatch,
+        common::RelDataDirection direction,
+        const std::vector<std::unique_ptr<arrow::Array>>& pkOffsets);
+    void checkViolationOfRelColumn(
+        common::RelDataDirection direction, arrow::Array* boundNodeOffsets);
+
+    static std::unique_ptr<arrow::PrimitiveArray> createArrowPrimitiveArray(
+        const std::shared_ptr<arrow::DataType>& type, const uint8_t* data, uint64_t length);
+
+protected:
+    std::shared_ptr<RelCopySharedState> sharedState;
+    common::CopyDescription copyDesc;
+    catalog::RelTableSchema* schema;
+    DirectedInMemRelData* fwdRelData;
+    DirectedInMemRelData* bwdRelData;
+    std::vector<common::LogicalType> pkColumnTypes; // src and dst node pk columns.
+    std::vector<PrimaryKeyIndex*> pkIndexes;
+};
+
+class RelListsCounterAndColumnCopier : public RelCopier {
+public:
+    RelListsCounterAndColumnCopier(std::shared_ptr<RelCopySharedState> sharedState,
+        const common::CopyDescription& copyDesc, catalog::RelTableSchema* schema,
+        DirectedInMemRelData* fwdRelData, DirectedInMemRelData* bwdRelData,
+        std::vector<common::LogicalType> pkColumnTypes, std::vector<PrimaryKeyIndex*> pkIndexes)
+        : RelCopier{std::move(sharedState), copyDesc, schema, fwdRelData, bwdRelData,
+              std::move(pkColumnTypes), std::move(pkIndexes)} {}
+
+protected:
```

Comment:
Since we make relCopier pure virtual, i think you need to define the clone method for `RelListsCounterAndColumnCopier `